Automattic/newspack-newsletters
# Fix: refactor constant contact provider

### All Submissions:






### Changes proposed in this Pull Request:

Implements our own [simplified SDK](https://github.com/Automattic/newspack-newsletters/blob/fix/refactor-constant-contact/includes/service-providers/constant_contact/class-newspack-newsletters-constant-contact-sdk.php) to handle the [maintained v3 API](https://v3.developer.constantcontact.com/api_guide/index.html) with OAuth integration.

Instead of storing a generated access token, the API credentials now store API's `key` and `secret`, used to manage access and refresh tokens. After the first configuration with the app's credentials, the editor sidebar is replaced with an Authorization button, which will prompt the OAuth flow, update the tokens seamlessly for the user and synchronize the post with the provider.

I struggled a little bit to implement the ability to synchronize the newsletter with the provider **after** the draft was created (either by the post being created with a different provider or before the tokens were available). I wasn't able to identify the root of this sync issue, but I identified that it was caused by the `cc_campaign_id` meta being available on the rest API. At some save cycle, the meta was being deleted and the post was desynced from the provider. Removing the meta from the API (see 2b1061b47fc63c37118105393709a26f7475142e) solved the issue. I figured it's not a problem, since we don't (and probably shouldn't) manage this data on the frontend.

#### Composer Autoloader update

To ensure that our Guzzle dependency does not conflict with an older version loaded from a different plugin, I'm using [this autoloader by Jetpack](https://github.com/Automattic/jetpack/tree/master/projects/packages/autoloader). According to its readme:

> This is a custom autoloader generator that uses a classmap to always load the latest version of a class.

Might be the case for further testing, but seems to work great here.



Closes #503.
Closes #511.
Closes #512.

### How to test the changes in this Pull Request:

1. Register a [Constant Contact Application](https://app.constantcontact.com/pages/dma/portal/)
2. Copy the API key and the generated secret
3. Update your Redirect URI setting to include: `https://newspack.local/wp-admin/index.php?cc_oauth2callback=*`
3. Make sure you have a Constant Contact account with updated information (or the API might fail sending)
4. Check out this branch and run `composer install && npm run build`
5. Go to Newsletters settings and set the provider to Constant Contact with your API credentials
6. Create a new newsletter and select a layout
7. Observe the "Authorize" message and button
8. Click on Authorize, login to your account if necessary, and Allow the app
9. Observe that the sidebar has been updated to synchronize with the provider
10. When it finishes retrieving the data, test by updating the content, sending test emails, and sending the newsletter.

In case your Constant Contact account is under review or suspended, you'll receive the following error notice while trying to send a test email or publishing:

<img width="1703" alt="Captura de Tela 2021-07-27 às 11 33 31" src="https://user-images.githubusercontent.com/820752/127173146-29bbe2b8-5c78-42fe-9e4b-6f76952564bb.png">

### Other information:
